go/database/mpt/verification_proof.go

```
@@ -0,0 +1,277 @@
+		if err := trie.VisitTrie(i, &visitor); err != nil {
+			observer.EndVerification(err)
+			return err
+		}
+		if interrupt.IsCancelled(ctx) {
+			return interrupt.ErrCanceled
+		}
+
+		observer.Progress(fmt.Sprintf("Collecting proof for block.. \t %d \t Accounts: %d", i, len(visitor.accounts)))
+		// collect all proofs for this block
+		rawProof := make(map[string]struct{})
+		for addr, data := range visitor.accounts {
+			if interrupt.IsCancelled(ctx) {
+				return interrupt.ErrCanceled
+			}
+
+			accountProof, err := trie.CreateWitnessProof(i, addr, maps.Keys(data.storage)...)
+			if err != nil {
+				observer.EndVerification(err)
+				return err
+			}
+			for _, element := range accountProof.GetElements() {
+				rawProof[element] = struct{}{}
+			}
+		}
+
+		rootHash, err := trie.GetHash(i)
+		if err != nil {
+			observer.EndVerification(err)
+			return err
+		}
+
+		observer.Progress(fmt.Sprintf("Recovering proof for block... \t %d", i))
+		// match proofs and database data
+		proof := CreateWitnessProofFromNodes(maps.Keys(rawProof))
+		observer.Progress(fmt.Sprintf("Verifying proof for block... \t %d", i))
+		if err := verifyProofState(ctx, rootHash, proof, visitor.accounts); err != nil {
+			observer.EndVerification(err)
+			return err
+		}
+	}
+
+	observer.EndVerification(nil)
+	return nil
+}
+
+// VerifyProofLiveTrie verifies the consistency of witness proofs for an live trie.
+// It reads the trie for the head block and loads all accounts and storage slots.
+// Then, it extracts witness proofs for the same accounts and storage slots.
+// Finally, the actual values from the trie are compared with the values extracted from the witness proofs.
+func VerifyProofLiveTrie(ctx context.Context, dir string, config MptConfig, observer VerificationObserver) (err error) {
+	trie, err := OpenFileLiveTrie(dir, config, NodeCacheConfig{})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = trie.Close()
```

err is not checked here